### PR TITLE
Ported writeRect to write 16bpp pixmaps from PaulStoffregen/ILI9341_t3 as well as DemoSauce examples

### PR DESCRIPTION
**Pull request by marcmerlin**
This is not my code, it was ported from
https://github.com/PaulStoffregen/ILI9341_t3

Tested on ESP32, allows writing 16bpp pixmaps defined in a C file.
image
tassie

### CODE REVIEW

File: Adafruit_ILI9341.cpp [hunk 1]
```diff
@@ -601,3 +601,30 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
     writeFillRect(x,y,w,h,color);
     endWrite();
 }
+
+// This code was ported/adapted from https://github.com/PaulStoffregen/ILI9341_t3
+// by Marc MERLIN
+void Adafruit_ILI9341::setAddr(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1) {
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: this method already exists ;) it's setAddressWindow

File: Adafruit_ILI9341.cpp [hunk 2]
```diff
@@ -601,3 +601,30 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
...
+	SPI_WRITE16(y0);   // YSTART
+	SPI_WRITE16(y1);   // YEND
+	writeCommand(ILI9341_RAMWR);
+}
+
+// See examples/pictureEmbed to use this.
+void Adafruit_ILI9341::writeRect(int16_t x, int16_t y, int16_t w, int16_t h, const uint16_t *pcolors) {
+        startWrite();
+	setAddr(x, y, x+w-1, y+h-1);
+	// setaddrwindow does not work if x!=1 (the picture gets offset) but setAddr works
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: I'm not sure I get this comment?

File: Adafruit_ILI9341.cpp [hunk 3]
```diff
@@ -601,3 +601,30 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
...
+	//setAddrWindow(x, y, x+w-1, y+h-1);
+	for(y=h; y>0; y--) {
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: this is equal to writePixels(pcolors, w*h)

File: Adafruit_ILI9341.cpp (hunk 2 above)
Comment (me-no-dev) [reply Adafruit_ILI9341.cpp]: read setAddressWindow ;) setAddr(x, y, x+w-1, y+h-1); == setAddessWindow(x,y,w,h)

File: Adafruit_ILI9341.cpp [hunk 4]
```diff
@@ -601,3 +601,30 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
...
+	SPI_WRITE16(y1);   // YEND
+	writeCommand(ILI9341_RAMWR);
+}
+
+// See examples/pictureEmbed to use this.
+void Adafruit_ILI9341::writeRect(int16_t x, int16_t y, int16_t w, int16_t h, const uint16_t *pcolors) {
+        startWrite();
+	setAddr(x, y, x+w-1, y+h-1);
+	// setaddrwindow does not work if x!=1 (the picture gets offset) but setAddr works
+	//setAddrWindow(x, y, x+w-1, y+h-1);
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: calling it wrong in this comment

File: Adafruit_ILI9341.cpp [hunk 5]
```diff
@@ -601,3 +601,30 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
...
+	for(y=h; y>0; y--) {
+		for(x=w; x>1; x--) {
+			SPI_WRITE16(*pcolors++);
+		}
+		SPI_WRITE16(*pcolors++);
+	}
+        endWrite();
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: since this method calls startWrite and endWrite, it should be called drawRect not writeRect

**Comment by me-no-dev** [comment]
```cpp
void Adafruit_ILI9341::drawRect(int16_t x, int16_t y, int16_t w, int16_t h, const uint16_t *pcolors) {
  startWrite();
  setAddressWindow(x,y,w,h);
  writePixels(pcolors, w*h);
  endWrite();
}
```

**Comment by marcmerlin** [comment]
Thank you for your review, clearly it was a bit too late last night for me to get rid of setAddr from the other code branch.
How does this cleaned up version look?

File: Adafruit_ILI9341.cpp (hunk 5 above)
Comment (marcmerlin) [reply Adafruit_ILI9341.cpp]: done.

**Comment by marcmerlin** [comment]
Ok, I used the time to finish porting the methods from https://github.com/PaulStoffregen/ILI9341_t3
that were missing for the included 10 screen animation demo, to work.
See a few screenshots and videos: https://goo.gl/photos/bMVqdiAxDkEppuN88
image
image

This should be a great stress test for your library. Sadly, on my board some of the demos crash, but this may due to the weak voltage regulator on my board.
Could you try integrating this patch, and running the demo on your board to see if all 10 demos work?
(reset 'DEBUG_ANIM = false' in DemoSauce)

File: Adafruit_ILI9341.cpp [hunk 6]
```diff
@@ -601,3 +601,25 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
     writeFillRect(x,y,w,h,color);
     endWrite();
 }
+
+// This code was ported/adapted from https://github.com/PaulStoffregen/ILI9341_t3
+// by Marc MERLIN. See examples/pictureEmbed to use this.
+void Adafruit_ILI9341::drawBitmap(int16_t x, int16_t y, int16_t w, int16_t h, const uint16_t *pcolors) {
+        startWrite();
+	setAddrWindow(x, y, w, h);
+	for(y=h; y>0; y--) {
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: as I said. those two for loops equal writePixels(pcolors, w*h)

File: Adafruit_ILI9341.cpp [hunk 7]
```diff
@@ -601,3 +601,25 @@ void Adafruit_ILI9341::fillRect(int16_t x, int16_t y, int16_t w, int16_t h,
...
+		for(x=w; x>1; x--) {
+			SPI_WRITE16(*pcolors++);
+		}
+		SPI_WRITE16(*pcolors++);
+	}
+        endWrite();
+}
+
+void Adafruit_ILI9341::setScroll(uint16_t offset)
```
Comment (me-no-dev) [inline Adafruit_ILI9341.cpp]: this is already in the library ;) scrollTo ;)

File: Adafruit_ILI9341.h [hunk 8]
```diff
@@ -155,6 +156,19 @@ class Adafruit_ILI9341 : public Adafruit_GFX {
         void      drawFastHLine(int16_t x, int16_t y, int16_t w, uint16_t color);
         void      fillRect(int16_t x, int16_t y, int16_t w, int16_t h, uint16_t color);
 
+	// Compat with ILI9341_T3 lib
```
Comment (me-no-dev) [inline Adafruit_ILI9341.h:159]: we need to figure out if the names for the methods make sense with the GFX library, or if we need to move them there...

**Comment by me-no-dev** [comment]
there are many many places where you call multiple draws in a loop of some sort. writing and not drawing could make things faster. See my other comments and look for math issues where the examples do not run.

File: Adafruit_ILI9341.cpp (hunk 7 above)
Comment (marcmerlin) [reply Adafruit_ILI9341.cpp]: First, thank you for having a look at all this, and you are of course totally correct. I got mislead by the fact that the ILI9341_VSCRSADD define was missing and assumed the method was too.
I've updated scrollTo to use ILI9341_VSCRSADD

File: Adafruit_ILI9341.h (hunk 8 above)
Comment (marcmerlin) [reply Adafruit_ILI9341.h:159]: Understood, this is your call. It's kind of a shame that Paul had a fork for so long and nothing got merged, and now we have diverging names.
If you want to remove the #defines and instead change the code to use the 'native' names, I'd understand.

File: Adafruit_ILI9341.cpp (hunk 6 above)
Comment (marcmerlin) [reply Adafruit_ILI9341.cpp]: My apologies, I missed that bit in your comment. Fixed, tested, and it works fine with writePixels

File: Adafruit_ILI9341.cpp (hunk 7 above)
Comment (marcmerlin) [inline Adafruit_ILI9341.cpp]: @me-no-dev does setScroll actually work for you?
I tried something like this, no joy:
  for (int16_t addr = 320; addr -=2; addr >= 0)
  {
    tft.scrollTo(addr);
    delay(100);
  }
}

**Comment by marcmerlin** [comment]
I tried to address all your comments (thanks again for the feedback), but about this one: "there are many many places where you call multiple draws in a loop of some sort. writing and not drawing could make things faster. See my other comments and look for math issues where the examples do not run."
Are you talking about code in demosauce? If so, that's not my code, I didn't look at it any longer than it took me to get it to run on my board.
If you mean something else I missed, please point me to what I should fix, and I'll have a look.

File: Adafruit_ILI9341.cpp (hunk 7 above)
Comment (marcmerlin) [reply Adafruit_ILI9341.cpp]: @me-no-dev so scrollTo wsa buggy and didn't work, but I just fixed it with this last uploaded commit.
I'm now able to display a bitmap and scroll it around.

**Comment by marcmerlin** [comment]
Ok, I fixed up the pictureEmbed.ino code a bit, and added a scrollTo example in it, which also validates the now fixed and working library code.
If you don't want to keep all the bitmaps or prune this down, feel free, I won't take offence :)

Also, I _think_ I'm done with the changes/improvement I needed on my side, so feel free to pull my changes, modify them in whichever way you see fit, including removing things you don't like if any, and I'll resync with you after that.
Thanks again for your time and looking at all this.

**Comment by marcmerlin** [comment]
Actually commit https://github.com/espressif/Adafruit_ILI9341/pull/1/commits/ae3517ddb6219c66aa919e7f174fa7a8c9f61ec0 isn't quite good.
Your suggestion to use writePixels the pixels are shifted in packs of 2 I think.
Try it out, but we may have to revert this one CL since the bitmap pixel order is reversed I think (need to do more testing, but I need to head out now)

**Comment by me-no-dev** [comment]
about writePixels, no it does the same exact thing that the for loops did. you increment the input address with each loop. And if you see the inner for loop, it counts down to 1 and not 0, but then it writes another pixel, which is the same exact thing.

As for scroll to, let me see what I have here (cuz I use it)... Yes damned... I must have translated it wrong when I was moving from C to C++

**Comment by marcmerlin** [comment]
@me-no-dev  I understand they should be the same, but they are not. 
The picture is visibly offset 2 pixels at a time when I use 
writePixels((uint16_t*) pcolors, w*h);
See 
image
When I revert to 
	for(y=h; y>0; y--) {
		for(x=w; x>1; x--) {
			SPI_WRITE16(*pcolors++);
		}
		SPI_WRITE16(*pcolors++);
	}
        endWrite();
the picture looks good again, see
image

So, clearly the picture looks damaged with 
writePixels((uint16_t*) pcolors, w*h);
Are you able to duplicate and see the difference too?

**Comment by me-no-dev** [comment]
I would rather get writePixels to work ;) if it is as you say, it's unusable

**Comment by marcmerlin** [comment]
@me-no-dev 
    size_t pixels = len;
    while(pixels--) SPI_WRITE16(*colors++);
worked fine, thanks for the suggestion.
Pull request updated with the new code using writePixels and I can confirm the bitmap is displayed properly now.

**Comment by marcmerlin** [comment]
I'm close to having my code ready for release, but having it depend on an unmerged pull request against the TFT lib, is non ideal :)
I know you're juggling a dozen things as is, but would you be ok merging the non demosauce part of the pull request? That part should be small and easy to review/add.
I may also have time to debug the non working demos in demosauce later, now that I can add extra power to my board to make sure it doesn't crash due to lack of current.

**Comment by me-no-dev** [comment]
@marcmerlin I have started working with the LCD and the code that I need to create (which includes images and whatnot) but have been delayed because the new wrover kits come with ST7789V instead of ILI9341 and I need to adjust the drivers. @ladyada has already merged the changes and we actually need to sync to that and then add the needed method in both drivers or if I can go with just one driver for AVR it will be even better.
Anyway diving into graphics and menus :)

**Comment by me-no-dev** [comment]
@marcmerlin so... I fixed writePixels ;) and talked to ladyada to fork the ili driver to WROVER_LCD and work on that instead. New WROVERs have different LCD chip ST7789V which is really similar to the ILI but has some differences in init and rotation procedures (so far).
The WROVER driver detects which LCD is attached and spins the proper code. I advise us to work on the driver instead and when we have something that is good for all, we can push it to the main ILI driver (and ST driver if created).
I got to write some code to display BMP images on the screen already :) 16 and 24 bit BMPs work. Now onto getting jpegs to display :) I think you are gong to like it ;) reads from the file system and displays.
I'm waiting for someone with Write privileges to Espressif's github to fork the ILI driver to the new name and we can get with it. Agreed?

**Comment by marcmerlin** [comment]
So, I think we can close this pull request, but you said you had a new drawBitmap and much more such methods you wrote, but I'm not seeing them in this code.
Can you make sure either
https://github.com/adafruit/Adafruit_ILI9341/pull/26
https://github.com/adafruit/Adafruit_ILI9341/pull/26/commits/63e77c01dcebb927c7ba681a4b62e6514a81ee04
https://github.com/adafruit/Adafruit_ILI9341/pull/26/commits/43fb49faf7f0a0ff1273bef33b3296932f825412
gets merged or push your new code that does a lot more?
Then we can close this pull request as obsolete :)

**Comment by marcmerlin** [comment]
this PR is now obsolete, but #2 replaces it by adding the one method that's still missing to draw bitmaps

**Pull request closed without merging**